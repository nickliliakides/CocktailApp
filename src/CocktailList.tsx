import { FC } from 'react';
import ICocktail from './Interfaces/ICocktail';
import './CocktailList.styles.scss';
import CocktailItem from './CocktailItem';
import Loader from './Loader';

interface IProps {
  cocktails: ICocktail[]
  loading: boolean
}

const CocktailList: FC<IProps> = ({ cocktails, loading }): JSX.Element => {
  return loading ? (
    <Loader />
  ) : (
    <div className='list-container'>
      <ul>
        {cocktails &&
          cocktails?.map((cocktail: ICocktail) => (
            <CocktailItem key={cocktail.id} cocktail={cocktail} />
          ))}
      </ul>
    </div>
  );
};

export default CocktailList;
