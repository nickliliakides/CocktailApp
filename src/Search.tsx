import React, { FC } from 'react';

interface IProps {
  handleSubmit: (e: React.FormEvent<HTMLElement>) => void
  handleChange: (e: React.ChangeEvent<HTMLInputElement>) => void
}

const Search: FC<IProps> = ({ handleSubmit, handleChange }): JSX.Element => {
  return (
    <form className='form' onSubmit={handleSubmit}>
      <div className='form-group'>
        <label htmlFor='search'>Find a cocktail by name</label>
        <div className='search'>
          <input
            className='form-control'
            id='search'
            name='search'
            type='search'
            placeholder='Please enter a cocktail name...'
            onChange={handleChange}
          />
          <button type='submit' className='btn btn-primary btn-lg'>
            Search
          </button>
        </div>
      </div>
    </form>
  );
};

export default Search;
