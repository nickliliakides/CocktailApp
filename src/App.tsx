import React, { FC, useState } from 'react';
import axios from 'axios';
import './App.styles.scss';
import ICocktail from './Interfaces/ICocktail';
import CocktailList from './CocktailList';
import Search from './Search';

const App: FC = (): JSX.Element => {
  const [searchField, setSearchField] = useState<string>('');
  const [cocktails, setCocktails] = useState<ICocktail[]>([]);
  const [loading, setLoading] = useState<boolean>(false);
  const apiHost = 'https://www.thecocktaildb.com/api/json/v1/1/search.php';

  const searchCocktails = async (url: string): Promise<any> => {
    try {
      setLoading(true);
      const {
        data: { drinks },
      } = await axios.get(url);
      if (drinks) {
        const newDrinks = drinks.filter((d: any, i: number): boolean => i < 5);
        const cocktails = newDrinks?.map((c: any) => {
          const ingredients: string[] = [];
          for (let i = 1; i < 16; i++) {
            if (c[`strIngredient${i}`]) {
              let ingredient = '';
              if (c[`strMeasure${i}`])
                ingredient = `${c[`strMeasure${i}`]}${c[`strIngredient${i}`]}`;
              else ingredient = `${c[`strIngredient${i}`]}`;
              ingredients.push(ingredient);
            }
          }
          return {
            id: c.idDrink,
            name: c.strDrink,
            img: c.strDrinkThumb,
            instructions: c.strInstructions,
            ingredients,
          };
        });
        setCocktails(cocktails);
        setLoading(false);
      } else setCocktails([]);
    } catch (error) {
      alert(
        `Error: ${error.response.status} - ${
          error.response.statusText
            ? error.response.statusText
            : 'Something went wrong!'
        }`
      );
      setLoading(false);
    }
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchField(e.currentTarget.value);
  };

  const handleSubmit = (e: React.FormEvent<HTMLElement>): void => {
    e.preventDefault();
    let url: string;
    if (searchField.trim().length) {
      url = `${apiHost}?s=${searchField.trim()}`;
      searchCocktails(url);
    } else {
      alert('Please enter a valid string');
    }
  };

  return (
    <>
      <Search handleChange={handleChange} handleSubmit={handleSubmit} />
      <CocktailList loading={loading} cocktails={cocktails} />
    </>
  );
};

export default App;
