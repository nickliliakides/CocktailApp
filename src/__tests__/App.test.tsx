import { shallow, ShallowWrapper } from 'enzyme';
import App from '../App';
import Search from '../Search';
import CocktailList from '../CocktailList';

let wrapper: ShallowWrapper;

describe('<App />', () => {
  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('should matches its snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the Search component', () => {
    expect(wrapper.find(Search).length).toEqual(1);
  });

  it('should render the CocktailList component', () => {
    expect(wrapper.find(CocktailList).length).toEqual(1);
  });
});
