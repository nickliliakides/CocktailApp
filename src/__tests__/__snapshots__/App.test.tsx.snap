// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should matches its snapshot 1`] = `
<Fragment>
  <Search
    handleChange={[Function]}
    handleSubmit={[Function]}
  />
  <CocktailList
    cocktails={Array []}
    loading={false}
  />
</Fragment>
`;
